temporalio/contrib/pydantic.py

```diff
@@ -0,0 +1,117 @@
+"""A data converter for Pydantic v2.
+
+To use, pass ``pydantic_data_converter`` as the ``data_converter`` argument to
+:py:class:`temporalio.client.Client`:
+
+.. code-block:: python
+
+    client = Client(
+        data_converter=pydantic_data_converter,
+        ...
+    )
+
+Pydantic v1 is not supported.
+"""
+
+from typing import Any, Optional, Type
+
+from pydantic import TypeAdapter, ValidationError
+from pydantic_core import to_json
+
+import temporalio.api.common.v1
+from temporalio.converter import (
+    CompositePayloadConverter,
+    DataConverter,
+    DefaultPayloadConverter,
+    EncodingPayloadConverter,
+    JSONPlainPayloadConverter,
+)
+
+# Note that in addition to the implementation in this module, _RestrictedProxy
+# implements __get_pydantic_core_schema__ so that pydantic unwraps proxied types.
+
+
+class PydanticJSONPlainPayloadConverter(EncodingPayloadConverter):
+    """Pydantic JSON payload converter.
+
+    Supports conversion of all types supported by Pydantic to and from JSON.
+
+    In addition to Pydantic models, these include all `json.dump`-able types,
+    various non-`json.dump`-able standard library types such as dataclasses,
+    types from the datetime module, sets, UUID, etc, and custom types composed
+    of any of these.
+
+    See https://docs.pydantic.dev/latest/api/standard_library_types/
+    """
+
+    @property
+    def encoding(self) -> str:
+        """See base class."""
+        return "json/plain"
+
+    def to_payload(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """See base class.
+
+        Uses :py:func:`pydantic_core.to_json` to serialize ``value` to JSON.
+
+        See
+        https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.to_json.
+        """
+        return temporalio.api.common.v1.Payload(
+            metadata={"encoding": self.encoding.encode()}, data=to_json(value)
+        )
+
+    def from_payload(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Any:
+        """See base class.
+
+        Uses :py:func:`pydantic.TypeAdapter.validate_json` to construct an
+        instance of the type specified by ``type_hint`` from the JSON payload.
+
+        See
+        https://docs.pydantic.dev/latest/api/type_adapter/#pydantic.type_adapter.TypeAdapter.validate_json.
+        """
+        _type_hint = type_hint if type_hint is not None else Any
+        try:
+            return TypeAdapter(_type_hint).validate_json(payload.data)
+        except ValidationError as err:
+            raise RuntimeError("Failed parsing") from err
```

Comment:
Per off-PR discussion, I wonder if we should just remove this wrapping (recognizing we probably should have never added it in the existing JSON converter)

Reply:
Agreed, done.